examples/step-70/step-70.cc

```diff
@@ -0,0 +1,1823 @@
+    // cells. On the other hand, by design we assign to the active process also
+    // the particles that were generated as support points of the locally owned
+    // parts of a non matching and arbitrary grid.
+    //
+    // The location of these particles is arbitrary, and may fall within a
+    // region that we don't have access to (i.e., a region of the fluid domain
+    // where cells are artificial). In order to understand who to send those
+    // particles to, we need to have a (rough) idea of how the fluid grid is
+    // distributed among processors.
+    //
+    // The following method computes a small collection of axis aligned
+    // bounding boxes that represent well the locally owned part of the fluid
+    // domain.
+    auto bounding_boxes_of_locally_owned_cells =
+      GridTools::compute_mesh_predicate_bounding_box(
+        fluid_tria, IteratorFilters::LocallyOwnedCell());
+
+    // These bounding boxes are then exchanged with all other processes, so
+    // that now every active process has a rough idea (based on bounding boxes
+    // that overlap with the locally owned part of the domain) of who they
+    // should send particles to.
+    auto global_bounding_boxes =
+      Utilities::MPI::all_gather(MPI_COMM_WORLD,
+                                 bounding_boxes_of_locally_owned_cells);
+
+
+    // Finally generate the particles from the support point of the
+    // tracer particles triangulation. This function call uses the
+    // global_bounding_boxes object we just constructed. At the end of this
+    // call, every particle will have been distributed to the correct process
+    // (i.e., the process that owns the cell where the particle lives).
+    Particles::Generators::dof_support_points(particles_dof_handler,
+                                              global_bounding_boxes,
+                                              tracer_particle_handler);
+
+    // As soon as we have initialized the particles in each process, we set
+    // their ownership to the current distribution. Since we will need this
+    // information to initialize a vector containing velocity information for
+    // each particle, we query the tracer_particle_handler for its locally
+    // relevant ids, and construct the indices that would be needed to store in
+    // a (parallel distributed) vector the position and velocity of all
+    // particles.
+    owned_tracer_particles =
+      tracer_particle_handler.locally_relevant_ids().tensor_product(
+        complete_index_set(spacedim));
+
+    // At the beginning of the simulation, all particles are in their original
+    // position. When particles move, they may jump to a part of the domain
+    // which is owned by another process. If this happens, the current process
+    // keeps formally "ownership" of the particles, but may need read access
+    // from the process where the particle has landed. We keep this information
```

Comment:
The communication only happens when visualization is requested or every time step?

Reply:
The communication happens every time step when the trilinos vector is used to carry out the interpolation. It was done this way to "mimic" a FSI interaction problem in which the solid parallel distributed triangulation would not be re-distributed every time step.